src/Adapter/File/Uploader/ProductImageUploader.php

```diff
@@ -0,0 +1,217 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\File\Uploader;
+
+use ErrorException;
+use Hook;
+use Image;
+use ImageManager;
+use PrestaShop\PrestaShop\Adapter\Image\Uploader\AbstractImageUploader;
+use PrestaShop\PrestaShop\Core\Configuration\UploadSizeConfigurationInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\CannotUnlinkImageException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\ImageConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\ImageException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\ImageNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\ImageUpdateException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\ImagePathFactoryInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\ProductImageUploaderInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\ValueObject\ImageId;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageOptimizationException;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageUploadException;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\MemoryLimitException;
+use PrestaShopException;
+use Symfony\Component\Mime\MimeTypes;
+
+final class ProductImageUploader extends AbstractImageUploader implements ProductImageUploaderInterface
+{
+    /**
+     * @var UploadSizeConfigurationInterface
+     */
+    private $uploadSizeConfiguration;
+
+    /**
+     * @var ImagePathFactoryInterface
+     */
+    private $productImagePathFactory;
+
+    /**
+     * @var array
+     */
+    private $contextShopIdsList;
+
+    /**
+     * @var int
+     */
+    private $contextShopId;
```

Comment:
Do you think we should create some structured data, useful, to carry things contextShopIdsList or $contextShopId ?

Example
```
<?php
class MultistoreContext
{
    private $shopId;
    private $groupdId;
    private $isAllShops;
    public function contextIsSingleShop()
    {
        return ($this->shopId !== null);
}
```
the idea would be to harmonize and provide util methods for everywhere
instead of copy-pasting always the same foreach loops in handler logic

Reply:
Need to check if MultistoreChecker(or smth like that) doesnt handle it already.